test/unit/sanitizer/_address_layer_filter.js

```diff
@@ -117,6 +117,26 @@ module.exports.tests.sanitize = function (test, common) {
     t.end();
   });
 
+  test('sanitize - exclude addresses when negative layers and sources are specified', (t) => {
+    // select all layers except venue to simulate value of clean.layers from targets sanitizer
+    const clean_layers = real_type_mapping.getCanonicalLayers().filter(layer => layer !== 'venue').sort();
+
+    let clean = { text: 'foo',
+      layers: clean_layers,
+      negative_layers: ['venue'],
+      positive_layers: [],
+      sources: ['openstreetmap', 'openaddresses','whosonfirst'],
+      negative_sources: ['geonames'],
+      positive_sources: []
+    };
+
+    const expected_layers = clean_layers.filter(layer => layer !== 'address').sort();
+
+    t.deepEqual(real_sanitizer.sanitize(null, clean), STD_MESSAGES);
+    t.deepEqual(clean.layers.sort(), expected_layers, 'layer list is reduced to exclude addresses');
+    t.end();
+  });
+
```

Comment:
I'm struggling to understand a few things about this unit test:
- The test is named "exclude addresses when negative layers and sources are specified", is this specific to the `address` layer?
- `clean_layers` filters out `venue` and sorts the array, then `expected_layers` does the same thing? Is this maybe to avoid sharing the same pointer as `clean_layers`?
- The selected `sources` are not being asserted.

Reply:
That's right, the entire `_address_layer_filter` file is centered around an optimization to filter out the address layer when two conditions are met:
- the input has certain parses, is below a certain number of tokens, etc
- the layers specified by the user are compatible with filtering out addresses. For example `layers=addresses` means we can't, but `layers=-venue` means we can.